csharp/src/Google.Protobuf/Reflection/DescriptorPool.cs

```diff
@@ -189,13 +186,38 @@ private static void ValidateSymbolName(IDescriptor descriptor)
             {
                 throw new DescriptorValidationException(descriptor, "Missing name.");
             }
-            if (!ValidationRegex.IsMatch(descriptor.Name))
+
+            if (!IsValidSymbolNameFormat(descriptor.Name))
             {
                 throw new DescriptorValidationException(descriptor,
                                                         "\"" + descriptor.Name + "\" is not a valid identifier.");
             }
         }
 
+        /// <summary>
+        /// Symbol name must start with a letter or underscore, and it can contain letters, numbers and underscores.
+        /// </summary>
+        private static bool IsValidSymbolNameFormat(string name)
+        {
+            if (!IsAsciiLetter(name[0]) && name[0] != '_')
+            {
+                return false;
+            }
+
+            for (int i = 1; i < name.Length; i++)
+            {
+                if (!IsAsciiLetter(name[i]) && !IsAsciiDigit(name[i]) && name[i] != '_')
+                {
+                    return false;
+                }
+            }
+
+            return true;
+
+            static bool IsAsciiLetter(char c) => (uint) ((c | 0x20) - 'a') <= 'z' - 'a';
```

Comment:
Possibly use `unchecked` to avoid throwing exceptions if this is ever compiled in a checked-by-default context and arithmetic overflows? (I believe we currently don't use checked-by-default, but it's not inconceivable that we might do so in the future, and the overflow here really *wouldn't* be a problem.)

Reply:
Are you sure? Feels a little YAGNI. And there are only 3 `unchecked` statements in the source today.

btw I copied `IsAsciiLetter` directly from https://github.com/dotnet/runtime/blame/faa1d0993afad9d751b2dca6bcfe356369ec6c08/src/libraries/System.Private.CoreLib/src/System/Char.cs#L236 (added in .NET 6 which Google.Protobuf doesn't target)